// prometheus-database-layer/database-server.js
//this is how we access the server by methods (locally);
var config = require("prometheus-config");
var simpleargs = require("simpleargs");
var connections = require('./mongo_connections.js');

var serverCache = connections(config.mongo_host, config.mongo_port); 

var Promise = require("promise");

var db_name = config.mongo_db_name;

var databaseServer = new function(){

	this.options = null;
	this.initialized = false;
	this.server = null;

	var that = this;

	this.isRemote = function(){
		if(this.options == null){
			return null;
		}else{
			return this.options.mode=="rest";
		}
	}

	this.init = function(options){
		this.options = options==undefined? {} : options;
		if(simpleargs(process.argv.slice(2))["l"]=="db"){
			this.options.mode="rest";
		}
		this.initialized=true;
	}

	this.query = function(collection, mode, query, options){
		console.log("I am in a query");
		options = options? options : {};
		return new Promise(function(resolve, reject){
			serverCache(db_name, collection, function(err, collection_object){
				if(err){
					reject(err);
					console.log("Error: ",err);
				}else{
					console.log("Mode", mode);
					switch(mode){
						case "find":
							console.log("I'm in find"); 
							collection_object[mode](query, options).toArray(function(err, val){
								console.log(val);
								if(err){
									reject(err)
								}else{
									resolve(val);
								}
							})
							break;
						case "insert":
						collection_object[mode](query, options, function(err, inserted){
							if(err){
								reject(err);
							}else{
								resolve(inserted);
							}
						})
						break;
						case "update":
							//`options` is new object value (entire document)
							console.log("update :D");
							collection_object[mode](query, options, function(err, inserted){
								if(err){
									reject(err);
								}else{
									resolve(inserted);
								}
							})
							break;
						}
					}
				})			
		})
	}
}

databaseServer.init();

module.exports = databaseServer;
