// prometheus-database-layer/database-layer.js
var Promise = require("bluebird");
var LayerManager = require("prometheus-layer-manager");


function allow_query(){
	if(LayerManager.isLocal()){
		return true;
	}else{
		if(LayerManager.layer=="web"){
			return false;
		}else{
			return true;
		}
	}
}

function db_server_setup_needed(){
	if(LayerManager.isLocal()){
		return true;
	}else{
		if(LayerManager.getLayer()=="db"){
			return true;
		}else{
			return false;
		}
	}
}


function rest_server_needed(){
	if(LayerManager.isLocal()){
		return false;
	}else{
		if(LayerManager.getLayer()=="db"){
			return true;
		}else{
			return false;
		}
	}
}

var server = null;

module.exports.initialized = false;

function init(){
	if(db_server_setup_needed()){
		server = require("./database-server.js");
		//server.init();
	}
	module.exports.initialized = true;
}

/*
module.exports.query = function(collection, mode, query){
	return new Promise(function(resolve, reject){
		return server.query(collection_name, mode, query, options);
	});
}
*/

module.exports.getServer = function(){
	if(!module.exports.initialized){
		init();
	}
	return server;
}


module.exports.init = function(){
	init();
}


function init_needed(){
	if(LayerManager.isLocal()){
		return true;
	}else{
		if(LayerManager.getLayer()=="web"){
			return false;
		}else{
			return true;
		}
	}
}

function rest_server_needed(){
	if(LayerManager.isLocal()){
		return false;
	}else{
		if(LayerManager.getLayer()=="db"){
			return true;
		}else{
			return false;
		}
	}
}


	if(rest_server_needed()){
		var rest_server = require("./database-rest-server.js");
		rest_server.start(function(){
			console.log("Database REST server started:", rest_server.info.uri);
		});
	}

	if(init_needed()){
		init();	
	}


var DatabaseAccess = require("prometheus-database-accessor");
