// prometheus-module-manager/module.js
var fs = require("fs");
var ModuleInfo = require("./module-info.js");
var ChipManager = require("prometheus-chip-manager");
var Channel = require("prometheus-channel");
var Service = require("prometheus-service");

var Module = new function(path_to_module){
	this.path = path_to_module;
	this.registred = false;
	this.executed = false;
	this.info = new ModuleInfo(path_to_module);
	this.id = this.info.id;
	this.body = require(path_to_module);
}

Module.prototype = new function(){

	this.generate_chip_action_method_name = function(chip_name, chip_type, action){
		return action + "_" + chip_type + "_" + "chip_name";
	}

	this.get_defined_chips_by_type = function(type){
		var ret = [];
		for(var i in this.info.defines){
			if(this.info.defines[i].indexOf(type)==0){
				ret.push(this.info.defines[i].split(".")[1]);
			}
		}
		return ret;
	}

	this.prepare_all_channel = function(){
		var channel_ids = this.get_defined_chips_by_type("channel");
		for(var i in channel_ids){
			var channel_object = new Channel();
			var prepare_function_name = this.generate_chip_action_method_name(channel_ids[i], "channel", "prepare");
			if(this.body[prepare_function_name]){
				this.body[prepare_function_name](channel_object);
			}
			ChipManager.registerChip(channel_object);
		}
	}

	this.prepare_all_service = function(){
		var service_ids = this.get_defined_chips_by_type("service");
		for(var i in service_ids){
			var service_object = new Service();
			var prepare_function_name = this.generate_chip_action_method_name(service_ids[i], "service", "prepare");
			if(this.body[prepare_function_name]){
				this.body[prepare_function_name](service_object);
			}
			ChipManager.registerChip(service_object);
		}
	}


	this.prepare = function(chip_types){
		for(var i in chip_types){
			this["prepare_all_"+chip_types[i]]();
		}
	}

	this.start = function(){

	}
}

module.exports = Module;
