// prometheus-user-manager/user-manager.js
var Promise = require("bluebird");
/*
function UserRepresentation(database_entry){
	this.database_entry = database_entry;
	if(!database_entry.body){
		console.log("database_entry missing:", database_entry.body);
	}
	console.log("DATABASE_ENTRY: ", database_entry);
}

UserRepresentation.prototype = new function(){
	var data = this.database_entry.body
	data.username = database_entry.username;
	data.userdata_id = database_entry.userdata_id;
}
*/

var UserManager = new function(){
var that = this;

	this.create_user = function(username, password, dispatcher){
		var user_data;
		return dispatcher.users_user_exists(username, dispatcher)
			.then(function(result){	
				if (!result){
					return dispatcher.resources_create("user_data", {})
						.then(function(user_data_resource){
							user_data = user_data_resource;
							return dispatcher.metadata_increment_variable("first_free_user_id", dispatcher);
						})
						.then(function(first_free_user_id){
							//console.log("user-manager", user_data_resource);
							var body = {
								user_id : first_free_user_id,
								username: username, 
								password: password, 
								userdata_id: user_data.id
							};	
							return dispatcher.database_query("users", "insert", body);
						})
				}else{
					console.log("==========rzucam Error============");
					return new Error("409.1");
				}
			})
	}

	this.user_exists = function(username, dispatcher){
		return new Promise(function(resolve, reject){
			dispatcher.database_query("users", "find", {username: username})
			.then(function(result){
				//console.log(result);
				resolve(result.length===1);
			});			
		})
	}



	this.password_match = function(username, password, dispatcher){
		username = username.toString();
		password = password.toString();
		return new Promise(function(resolve, reject){
			//console.log("username: ", username);
			//console.log("password: ", password);
			dispatcher.database_query("users", "find", {username: username, password: password})
			.then(function(result){
				resolve(result);
			})			
		})
	}

	this.get_all_users = function(dispatcher){
		return dispatcher.database_query("users", "find")
	}

	this.get_user_data = function(userdata_id, dispatcher){
		return new Promise(function(resolve, reject){
			dispatcher.database_query("users", "find", {userdata_id: parseInt(userdata_id)})
				.then(function(userdata){
					delete userdata[0]["_id"];
					resolve(userdata[0]);
				})			
		})
	}

	this.update_user_data = function(username, new_user_data, dispatcher){
		return new Promise (function(resolve, reject){		
			dispatcher.users_user_exists(username, dispatcher)
			.then(function(exists){
				if(exists){
					dispatcher.database_query("users", "update", {username: username}, {"$set":new_user_data})
					.then(function(documents){
						resolve();
					})
				}else{
					reject(new Error("404"))
				}
			})
		})
	}

	this.delete_user = function(username, dispatcher){
 		return new Promise(function(resolve, reject){ 			
 			dispatcher.database_query("users", "delete", {username: username})
 			.then(function(data){
				resolve(data);
			}).catch(function(e){
	 			reject(e);
	 		});			
 		});
 	}



}

module.exports = UserManager;
