// prometheus-resource/resource-manager.js
var Promise = require("bluebird");

var DatabaseInterface = require("prometheus-database")["interface"];
var metadata = require('prometheus-metadata-manager');
var ResourceTypeManager = require("prometheus-resource").ResourceTypeManager;
var Resource = require("prometheus-resource");

/**
 * Manages resources in the database
 * @class 
 */
 var ResourceManager = new function(){

 	this.newResource = function(type_name, body){

 		console.log("new resource:", body);

 		ResourceTypeManager.enforce_correct_type_name(type_name);

 		var resource_type_object = ResourceTypeManager.getByName(type_name);

 		return new Promise(function(resolve, reject){
 			resource_type_object.validateFieldValues(body)
 				.then(
 					function(){
	 					return resource_type_object.encodeFieldValues(body); 						
 					}
 				).then(
	 				function(encoded_values){
	 					console.log("encoded_values:", encoded_values);
	 					metadata.increment("first_free_id", function(newID) {
	 						DatabaseInterface.query("resources", "insert", {prometheus_id: newID, type: type_name, body: encoded_values}, {}).then(function(data){
	 							var database_entry = data[0];
	 							var resource = new Resource(database_entry);
	 							resolve(resource);
	 						});
	 					});				
	 				}
 				).catch(function(e){
 					reject(e);
 				});			
 			})

 	}

	/**
	 * Calls callback function with `true` if a resource with given ID exists in the database. Otherwise, 
	 * @param  {Number}   prometheus_id 	the id to test for
	 * @param  {string}   [type] 			type of the resource. If provided, callback value will be true only if a resource with a given id AND of a provided type exists
	 * @param  {ResourceManager~idExistsCallback} callback
	 */
	 this.idExists = function(prometheus_id, type_name, callback){
	 	if(arguments.length==2){
	 		callback = arguments[1];
	 		type_name = undefined;
	 	}
	 	var query = {
	 		prometheus_id: parseInt(prometheus_id)
	 	}
	 	if(type_name){
	 		ResourceTypeManager.enforce_correct_type_name(type_name);
	 		query.type = type_name;
	 	}
	 	DatabaseInterface.query("resources", "find", query, {}).then(function(response){
	 		console.log("idExists database callback:", response);
	 		if(response.length===0){
	 			callback(false);
	 		}else{
	 			callback(true);
	 		}
	 	})
	 }

	/**
	 * Callback for idExists
	 * @callback ResourceManager~idExistsCallback
	 * @param {Boolean} exists - true if a resource with given id exists
	 */

	 this.getResourceByID = function(resource_id, callback){
	 	var promise = DatabaseInterface.query("resources", "find", { prometheus_id: resource_id }, {})
	 	if(callback){
	 		promise.then(function(response){
	 			var resource = new Resource(response);
	 			callback(resource);
	 		})
	 	}else{
	 		return promise;
	 	}
	 }

	 this.getResourcesByType = function(type_name, params, callback) {
	 	ResourceTypeManager.enforce_correct_type_name(type_name);
	 	type_name = type_name.toString();
	 	DatabaseInterface.query("resources", "find", { type: type_name }, {}, params).then(function(response) {
	 		var ret = response.map(function(database_entry){
	 			return new Resource(database_entry);
	 		})
	 		callback(ret);
	 	})
	 }
	}

	module.exports = ResourceManager;
