// prometheus-dispatcher-wrapper/dispatcher-wrapper.js
var Set = require("Set");

function DispatcherWrapper(real_dispatcher, allowed_chip_longids){

	this.real_dispatcher = real_dispatcher;


	this.allowed_chip_longids = new Set();
	this.allowed_chip_longids.addAll(allowed_chip_longids);

}

DispatcherWrapper.prototype = new function(){

	var that = this;

	this.resources_list_by_type = function(type_name){
		if(this.can_access("resource_type." + type_name)){
			return this.real_dispatcher.resources_list_by_type(type_name);
		}else{
			throw new Error("cannot access chip: resource_type."+type_name);
		}
	}

	this.resources_create = function(type_name, body){
		if(this.can_access("resource_type." + type_name)){
			return this.real_dispatcher.resources_create.apply(this.real_dispatcher, arguments);
		}
	}

	this.can_access = function(requested_chip_longid){
		if (this.allowed_chip_longids.has("resource_type.*") && requested_chip_longid.indexOf("resource_type") === 0) {
			return true;
		} else {
			return this.allowed_chip_longids.has(requested_chip_longid);
		}
	}

}

module.exports = DispatcherWrapper;
