// prometheus-database-accessor/database-accessor.js
var Promise = require("promise");
var request = require("request");
var queryString = require("query-string");

var LayerManager = require("prometheus-layer-manager");
var config = require("prometheus-config");

var db_rest_api_client = new function(){

	this.init = function(){
		//server_configuration = LayerManager.getDBLayerLocation();
	}

	function generateUrl(){
		var remote_settings = config.db_layer_config;
		return "http://" + remote_settings.host + ":" + remote_settings.port;
	}

	this.query = function(collection_name, mode, query, options){
		console.log("I am a DB API REST client and I did receive a request to send a request");
		return new Promise(function(resolve, reject){
			if(mode=="find"){
				console.log("I am in FIND mode");
				var url = generateUrl() + "/api/rest/v1/" + collection_name + "/";
				var query_string = queryString.stringify({
				  mode: mode,
				  query: JSON.stringify(query),
				  options: JSON.stringify(options)
				});

				var entire_url = url+"?"+query_string
				console.log("going to call GET ", entire_url);
				request.get(entire_url, function(err, res, body){
					console.log("REQUEST RESPONSE BODY:", body);
					console.log("error:", err)
					if(err){
						throw new Error(err);
						reject(err);
					}else{
						console.log("GOT RESPONSE!", body);
						resolve(body);
					}
				})
			}	
			
			if(mode=="insert"){
				var url = generateUrl() + "/api/rest/v1/" + collection_name + "/";
				var query_string = queryString.stringify({
				  mode: mode,
				  query: JSON.stringify(query),
				  options: JSON.stringify(options)
				});

				console.log("going to call POST ", url);
				request.post(url, function(err, res, body){
					console.log("REQUEST RESPONSE BODY:", body);
					console.log("error:", err)
					if(err){
						throw new Error(err);
						reject(err);
					}else{
						console.log("GOT RESPONSE!", body);
						resolve(body);
					}
				})
			}	



			if(mode=="update"){
				
			}
		})
	}
}

var db_lpc_api_client = new function(){

	var server = null;

	this.init = function(){
		console.log("initializing local database accessor");
		server = require("prometheus-database-layer").getServer();
		this.query = server.query;
	}

}


function lpc_api_needed(){
	if(LayerManager.isLocal()){
		return true;
	}else{
		if(LayerManager.getLayer()=="web"){
			return false
		}else{
			return true;
		}
	}
}


function rest_api_needed(){
	if(LayerManager.isLocal()){
		return false;
	}else{
		if(LayerManager.getLayer()=="biz"){
			return true;
		}else{
			return false;
		}
	}
}


if(rest_api_needed()){
	module.exports = db_rest_api_client;	
}else{
	if(lpc_api_needed()){
		db_lpc_api_client.init();
		module.exports = db_lpc_api_client;		
	}
}
