// prometheus-resource-type/resource-type.js
var Promise = require("bluebird");
var ResourceTypeField = require("./resource-type-field.js");
var merge = require("merge");

var ResourceType = function(name){
	this.id = name;
	this.name = name;
	this.fields = {};
}

ResourceType.prototype = new function(){

	this.add_field = function(field_declaration){
		var field_object = new ResourceTypeField(field_declaration);
		var field_name = field_object.name;
		if(!this.fields[field_name]){
			this.fields[field_name] = field_object;
		}
	}

	this.add_fields = function(field_declarations_array){
		for(var i in field_declarations_array){
			var declaration = field_declarations_array[i];
			this.add_field(declaration);
		}
	}

	this.validate_field_values = function(values){
		return new Promise(function(resolve, reject){
			for(var i in this.fields){
				var field = this.fields[i];
				if(field.required && !value[field.name]){
					reject(new Error("required field missing:" + field.name))
					return;
				}
			}
			resolve();
		})
	}

	this.encode_field_values = function(body){
		console.log("###########encoding body:", body);
		var promises = [];
		for(var field_name in body){
			var current_value = body[field_name];
			promises.push(this.fields[field_name].encode_value(current_value, true));
		}
		return Promise.all(promises).then(function(responses){
			return new Promise(function(resolve, reject){
				resolve(merge.apply(merge, responses));
			})
		});
	}

}

module.exports = ResourceType;
