// prometheus-layer-manager/interfaces/db-interfaces.js
/**
 * Provides access to the database
 * @module database-accessor
 */
var DatabaseDirectAccess = require("prometheus-database").direct_access;
var Promise = require("bluebird");
var request = require("request");
var queryString = require("query-string");

var LayerManager = require("prometheus-layer-manager");
var config = require("prometheus-config");
var PrometheusError = require("prometheus-error");

var db_rest_api_client = new function(){

	this.init = function(){
		//server_configuration = LayerManager.getDBLayerLocation();
	}

	function generateUrl(){
		var remote_settings = config.db_layer_config;
		return "http://" + remote_settings.host + ":" + remote_settings.port;
	}

	this.query = function(collection_name, mode, query, options, output_options){
		output_options = output_options || {};
		return new Promise(function(resolve, reject){
			if(mode=="find"){
				var url = generateUrl() + "/api/rest/v1/" + collection_name + "/";

				var query_string = queryString.stringify({
				  mode: mode,
				  query: JSON.stringify(query),
				  options: JSON.stringify(options),
				  output_options: JSON.stringify(output_options)
				});

				var entire_url = url+"?"+query_string

				request.get(entire_url, function(err, res, body){
					if(err){
						throw new PrometheusError(err, "find_error", "database-accessor");
						reject(err);
					}else{
						resolve(JSON.parse(body));
					}
				})
			}	
			
			if(mode=="insert"){
				var url = generateUrl() + "/api/rest/v1/" + collection_name + "/";
				var payload = {
				  mode: "insert",
				  query: query,
				  options: options
				};

				var payload_string = JSON.stringify(payload);	


				request.post(url, {form:{json_encoded_payload:payload_string}}, function(err, res, body){
					if(err){
						throw new PrometheusError(err, "insert_error", "database-accessor");
						reject(err);
					}else{
						resolve(JSON.parse(body));
					}
				})
			}	



			if(mode=="update"){
				var url = generateUrl() + "/api/rest/v1/" + collection_name + "/";
				var payload = {
				  mode: "update",
				  query: query,
				  options: options
				};

				var payload_string = JSON.stringify(payload);


				request.put(url, {form:{json_encoded_payload:payload_string}}, function(err, res, body){
					if(err){
						throw new PrometheusError(err, "update_error", "database-accessor");
						reject(err);
					}else{
						resolve(JSON.parse(body));
					}
				})
			}
		})
	}
}

module.exports.distributed = db_rest_api_client;
module.exports.local = DatabaseDirectAccess;
