// prometheus-chip-manager/chip-manager.js
var fs = require("fs");
var path = require('path')
var Set = require('Set');
//var Channel = require('prometheus-channel')

var channels = {};
var channel_info = {};

var services = {};
var service_info = {};

var registred_chips = new Set();

var Core = new function(){

	this.registerChannel = function(id, info, body){
		if(arguments.length==3){
			service_info[id] = info;
			channels[id] = body;			
		}else{
			var channel_id = arguments[0].id;
			channels[channel_id] = arguments[0];
		}
	}

	this.registerService = function(id, info, body){
		if(arguments.length==3){
			service_info[id] = info;
			services[id] = body;			
		}else{
			service_id = arguments[0].id;
			services[service_id] = arguments[0];
		}
	}
	
	this.get_chip_by_long_id = function(longid){
		var type = longid.split(".")[0];
		var name = longid.split(".")[1];
		switch(type){
			case "service":
				return this.getService(name);
				break;
			case "channel":
				return this.getChannel(name);
				break;
			default:
				return null;
				brak;
		}
	}

	this.chipIsRegistred = function(chip_name){
		return registred_chips.has(chip_name);
	}

	this.getChannel = function(id){
		return channels[id];
	}

	this.getService = function(id){
		return services[id] || null;
	}

	this.getServices = function(){
		return services;
	}
}


module.exports = Core;
//!!important! export Core before importning Modules (circular dependency fail)
Modules = require("prometheus-module-manager");
