// prometheus-layer-manager/layer-manager.js
var simpleargs    = require("simpleargs");
var ModuleManager   = require("prometheus-module-manager").ModuleManager;
var BizInterfaces = require("./interfaces/biz-interfaces.js");
var DBInterfaces  = require("./interfaces/db-interfaces.js");


var LayerManager = new function(){

	this.mode = null;
	var stateLocal = null;
	var l_argument = null;
	var layerName = null;
	var isValid = null;

	this.isLocal = function(){
		return stateLocal;
	}

	this.getLayer = function(){
		return layerName;
	}
	this.isValid = function(){
		return isValid;
	}

	this.biz_interface = null;
	this.db_interface = null;

	this.init = function(){
		this.mode = l_argument || "local";
		if(this.isLocal()){
			//local
			this.biz_interface = BizInterfaces.local;
			this.biz_interface.init();
			this.db_interface = DBInterfaces.local;
			this.db_interface.init();
			ModuleManager.execute_modules(["services", "channels"]);
		}else{
			//distributed
			this.biz_interface = BizInterfaces.distributed;
			this.db_interface = DBInterfaces.distributed;
			switch(this.getLayer()){
				case "db":
				db_layer.init();
				break;
				case "biz":
				biz_layer.init();
				ModuleManager.execute_modules(["services"]);
				break;
				case "web":
				web_layer.init();
				ModuleManager.execute_modules(["channels"]);
				break;
			}
		}

		return false;
	}

	var checkLocal = function(){
		l_argument = simpleargs(process.argv.slice(2))["l"];
		stateLocal = l_argument==="local" || l_argument===null || l_argument==undefined;
	}
	var checkLayer = function(){
		if(!l_argument){
			layerName =  null;
		}else{
			layerName = l_argument;
		}
	}
	var checkValidation = function(){
		if(process.argv.length > 3) {
			var arg = process.argv[3];
			if(typeof(arg) === 'undefined' || arg === null || (arg!="web" && arg!="biz" && arg != "db")) {
				isValid = false;
			}

		}
		isValid = true;
	}

	checkLocal();
	checkLayer();
	checkValidation();
}

module.exports = LayerManager;
