// prometheus-layer-manager/layer-manager.js
var simpleargs = require("simpleargs");
var Modules = require("prometheus-module-manager");

var LayerManager = new function(){

	this.l_argument = null;
	this.isValid = null;
	this.mode = null;
	this.isLocal = function(){
		var l_a = simpleargs(process.argv.slice(2))["l"]
		this.l_argument =  l_a==="local" || l_a===null || l_a==undefined;
	}

	this.getLayer = function(){
		var l_argument = simpleargs(process.argv.slice(2))["l"]
		if(!l_argument){
			return null;
		}else{
			return l_argument;
		}
	}
	this.validate = function(){
		var arg = process.argv[3];
		if(process.argv.length>3){
			if(typeof(arg) === 'undefined' || arg === null || (arg!="web" && arg!="biz" && arg != "db")) {
				this.isValid = false;
			}
		}
		this.isValid = true;
	}

	this.init = function(){
		this.mode = this.l_argument || "local";
		if(this.l_argument){
			for(var i in layers){
				layers[i].init();
			}
			Modules.execute_modules(["services", "channels"]);
		}else{
			switch(this.getLayer()){
				case "db":
				db_layer.init();
				break;
				case "biz":
				biz_layer.init();
				Modules.execute_modules(["services"]);
				break;
				case "web":
				web_layer.init();
				Modules.execute_modules(["channels"]);
				break;
			}
		}

		return false;
	}

}

module.exports = LayerManager;

var web_layer = require("prometheus-web-layer");
var biz_layer = require("prometheus-business-layer");
var db_layer = require("prometheus-database-layer");

var layers = {
	"db": db_layer,
	"web": web_layer,
	"biz": biz_layer,
}
