// prometheus-resource-type-field/resource-type-field.js
var Promise = require("bluebird");
var FieldTypeManager = require("prometheus-field-type-manager");

/**
 * A field in a resource type
 * @class
 * @param {string} name
 * @param {object} options
 */
function ResourceTypeField(name, options){
	this.name = name;
	this.type_name = options.type;
	var type_constructor = FieldTypeManager.getFieldTypeConstructorByName(options.type);
	this.type = new type_constructor(options.name);
	this.required = options.required || false;
	this.derived = options.derived || false;
}

ResourceTypeField.prototype = new function(){
	/**
	 * Shorthand for ResourceTypeField.type.isProperValue
	 * @alias ResourceTypeField#isProperValue
	 * @param  {object}  value
	 * @return {Promise}
	 */
	this.isProperValue = function(value){
		return this.type.isProperValue(value);
	}

	/**
	 * Encodes a value for this field so it can be stored safey in database. Reverse of @link ResourceTypeField#decodeValue
	 * @alias ResourceTypeField#encodeValue
	 * @param  {any} value
	 * @param  {Boolean} as_hashmap
	 * @return {Promise}
	 */
	this.encodeValue = function(value, as_hashmap){
		var that = this;
		return this.type.encode(value).then(function(encoded_value){
			var ret_promise = new Promise(function(resolve, reject){
				if(as_hashmap){
					var ret = {};
					ret[that.name] = encoded_value;
					resolve(ret);
				}else{
					resolve(encoded_value);
				}
			})
			return ret_promise;
		});
	}

	/**
	 * @alias ResourceTypeField#decodeValue
	 * @todo Zaimplementować tę funkcję
	 */
	 this.decodeValue = function(){

	 }
}

module.exports = ResourceTypeField;
